### PR TITLE
add hgraph param

### PR DESCRIPTION
**Pull request by LHT129**
issue: #40 
- introduce inner param for inner indexes
- map the external param to inner
- add format_map for long string to help reading

### CODE REVIEW

File: src/index/hgraph_zparameters.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+void
+HgraphParameters::ParseStringParam(const std::string& str) {
+    auto json_obj = nlohmann::json::parse(str);
+    for (const auto& [key, value] : json_obj.items()) {
+        this->CheckAndSetKeyValue(key, value);
+    }
+}
+
+void
+HgraphParameters::CheckAndSetKeyValue(const std::string& key, nlohmann::json& value) {
```
Comment (jiaweizone) [inline src/index/hgraph_zparameters.cpp]: nlohmann::json can define to `JsonType` ? like https://github.com/alipay/vsag/pull/96

File: src/index/hgraph_zparameters.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        for (const auto& str : vec) {
+            json = json[str];
+        }
+        json = value;
+    } else {
+        // TODO(LHT): Error logger
+    }
+}
+
+const std::unordered_map<std::string, std::vector<std::string>> HgraphParameters::EXTERNAL_MAPPING =
```
Comment (jiaweizone) [inline src/index/hgraph_zparameters.cpp]: move to L25 ?
let the `variables definition` precede `functions definition`

File: src/index/hgraph_zparameters.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+#pragma once
+
+#include <nlohmann/json.hpp>
+#include <string>
+
+#include "index_common_param.h"
+
+namespace vsag {
+class HgraphParameters {
```
Comment (jiaweizone) [inline src/index/hgraph_zparameters.h]: minor: HGraphParameters ?

File: src/index/hgraph_zparameters.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    nlohmann::json
+    GetJson() {
+        this->sync_json_by_string();
+        return this->json_;
+    }
+
+private:
+    void
+    sync_json_by_string() {
```
Comment (jiaweizone) [inline src/index/hgraph_zparameters.h]: refresh_json ?

File: src/index/hgraph_zparameters.h [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        this->json_ = nlohmann::json::parse(str_);
+    }
+
+    void
+    sync_string_by_json() {
```
Comment (jiaweizone) [inline src/index/hgraph_zparameters.h]: refresh_string ?

File: src/utils.cpp [hunk 6]
```diff
@@ -83,4 +83,18 @@ WindowResultQueue::GetAvgResult() const {
...
 }
 
+std::string
+format_map(const std::string& str, const std::unordered_map<std::string, std::string>& mappings) {
+    std::string result = str;
+
+    for (const auto& [key, value] : mappings) {
+        size_t pos = result.find("{" + key + "}");
+        while (pos != std::string::npos) {
+            result.replace(pos, key.length() + 2, value);
```
Comment (jiaweizone) [inline src/utils.cpp:93]:
```C++
result = result.replace(pos, key.length() + 2, value)
```

File: src/index/hgraph_zparameters.cpp (hunk 1 above)
Comment (LHT129) [reply src/index/hgraph_zparameters.cpp]: we will alias all of this json type in another PR

File: src/index/hgraph_zparameters.h (hunk 5 above)
Comment (LHT129) [reply src/index/hgraph_zparameters.h]: ditto

File: src/index/hgraph_zparameters.h (hunk 4 above)
Comment (LHT129) [reply src/index/hgraph_zparameters.h]: I think the refresh_json_by_string is better

File: src/index/hgraph_zparameters.h (hunk 3 above)
Comment (LHT129) [reply src/index/hgraph_zparameters.h]: done

File: src/index/hgraph_zparameters.cpp (hunk 2 above)
Comment (LHT129) [reply src/index/hgraph_zparameters.cpp]: done

File: src/utils.cpp (hunk 6 above)
Comment (LHT129) [reply src/utils.cpp:93]: https://en.cppreference.com/w/cpp/string/basic_string/replace maybe no need care return value

File: src/index/hgraph_zparameters.cpp (hunk 1 above)
Comment (jiaweizone) [reply src/index/hgraph_zparameters.cpp]: OK

File: src/index/hgraph_zparameters.h (hunk 4 above)
Comment (jiaweizone) [reply src/index/hgraph_zparameters.h]: OK

File: src/utils.cpp (hunk 6 above)
Comment (jiaweizone) [reply src/utils.cpp:93]: OK，I known C++ string is mutable :)

**Review by jiaweizone — APPROVED** [review]
LGTM

**Review by wxyucs — APPROVED** [review]
lgtm

**Pull request merged**